### PR TITLE
add Azure Pipelines Scaler

### PR DESCRIPTION
**Pull request by troydn**
Add a new Azure Pipelines Scaler

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] Tests have been added
~~- [ ] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*~~
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)* (https://github.com/kedacore/keda-docs/pull/411)
- [x] Changelog has been updated

Relates to #1705

### CODE REVIEW

**Comment by tomkerkhove** [comment]
Thanks a ton @troydn, this is great!

File: pkg/scalers/azure_pipelines_scaler.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+	}
+
+	return count, err
+}
+
+func (s *azurePipelinesScaler) GetMetricSpecForScaling() []v2beta2.MetricSpec {
+	targetPipelinesQueueLengthQty := resource.NewQuantity(int64(s.metadata.targetPipelinesQueueLength), resource.DecimalSI)
+	externalMetric := &v2beta2.ExternalMetricSource{
+		Metric: v2beta2.MetricIdentifier{
+			Name: kedautil.NormalizeString(fmt.Sprintf("%s-%s", "azure-pipelines-queue", s.metadata.poolID)),
```
Comment (zroubalik) [inline pkg/scalers/azure_pipelines_scaler.go:161]: `s.metadata.poolID` - is the poolID unique on Azure org level or is it unique across the whole Azure platform? We need to have the generated metricName unique enough, so if there are two triggers in the same ScaledObject, we can distinguish them by the metricName. 

So here comes the hypotetical example, can we have two Azure Pipelines triggers in the same ScaledObject that both have the same `poolID` but are in different Azure org?  -> Which will result in the same `metricName` for both triggers, but practically they are pointing to different Azure Pipelines.

**Review by zroubalik — COMMENTED** [review]
Looking good, minor comment!

File: pkg/scalers/azure_pipelines_scaler.go (hunk 1 above)
Comment (troydn) [reply pkg/scalers/azure_pipelines_scaler.go:161]: You are correct, this should be more unique. I will change this so the organization name is included as well.

**Review by zroubalik — APPROVED** [review]
LGTM, thanks @troydn 

Would be nice if you could add e2e test for this scaler.
https://github.com/kedacore/keda/tree/main/tests

**Comment by troydn** [comment]
> LGTM, thanks @troydn
> 
> Would be nice if you could add e2e test for this scaler.
> https://github.com/kedacore/keda/tree/main/tests

I can add e2e tests for this scaler but this would require an Azure DevOps organization that can be used for the e2e tests. Should I create one and provide the details/access or would you rather provide one?

**Comment by tomkerkhove** [comment]
I can set one up, what do you need? Org id and PAT?

**Comment by tomkerkhove** [comment]
And what PAT permissions?

**Comment by zroubalik** [comment]
@tomkerkhove great, just to be sure, we will need something pernament once the tests are ready. Mabye @jeffhollan or @ahmelsayed could help with that?

**Comment by tomkerkhove** [comment]
They are free so should be OK!

**Comment by tomkerkhove** [comment]
Org created as per https://github.com/kedacore/governance/issues/20 and is up on

**Comment by troydn** [comment]
> Org created as per [kedacore/governance#20](https://github.com/kedacore/governance/issues/20) and is up on https://dev.azure.com/kubernetes-event-driven-autoscaling/Kubernetes%20Event-driven%20Autoscaling%20(KEDA)

Thanks!
Should the e2e test also manage the creation of the build definition and the agent pool on Azure DevOps? It could also be a one time setup where an empty agent pool is created and a simple build definition is created that sleeps for 2 minutes.

**Comment by tomkerkhove** [comment]
I'm happy to set it up, I'll create an empty pool and pipeline that has a 2 min delay in it.

What else do you need in terms of config and what PAT permissions are required?

**Comment by troydn** [comment]
> I'm happy to set it up, I'll create an empty pool and pipeline that has a 2 min delay in it.
> 
> What else do you need in terms of config and what PAT permissions are required?

I finished the e2e test. But I'm not sure what image should be used for the Azure DevOps Agent. For now, I used one on my docker registry. But I think this should be replaced by one on a registry managed by KEDA.

The test requires an Azure DevOps environment with the following config:
- Agent pool created
- A pipeline configured to run on the created agent pool with "sleep 2m" as build step
- A PAT with the following permissions:
  - Agent Pools : Read & manage
  - Build: Read & execute

**Comment by tomkerkhove** [comment]
> > I'm happy to set it up, I'll create an empty pool and pipeline that has a 2 min delay in it.
> > What else do you need in terms of config and what PAT permissions are required?
> 
> I finished the e2e test. But I'm not sure what image should be used for the Azure DevOps Agent. For now, I used one on my docker registry. But I think this should be replaced by one on a registry managed by KEDA.

We host all test images on https://github.com/kedacore/test-tools, are you OK in adding a container image there? 

> The test requires an Azure DevOps environment with the following config:
> 
> * Agent pool created
> * A pipeline configured to run on the created agent pool with "sleep 2m" as build step
> * A PAT with the following permissions:
>   
>   * Agent Pools : Read & manage
>   * Build: Read & execute

I'll get this setup tomorrow.

**Comment by tomkerkhove** [comment]
@troydn I've added these GitHub secrets for you:
image

Pipeline is up on https://dev.azure.com/kubernetes-event-driven-autoscaling/Kubernetes%20Event-driven%20Autoscaling%20(KEDA)/_build?definitionId=1 with PR open on https://github.com/kedacore/test-tools/pull/11

**Comment by troydn** [comment]
@tomkerkhove thanks for your help on this. I think the e2e test is ready now.

**Review by tomkerkhove — APPROVED** [review]
LGTM, can you just fix the merge conflict please?

**Comment by tomkerkhove** [comment]
Thank you @troydn and let me know if you are interested for the blog post! (kerkhove.tom@gmail.com)

**Comment by mo-saeed** [comment]
@troydn I have very strange behaviour when I am using `scalerJob` which works fine but always create useless job with no pods running

so I have 1 pending pipeline job and here's the result from the kubernetes namespace 

```
 kubectl get jobs -n azure-devops
NAME                         COMPLETIONS   DURATION   AGE
azure-devops-runner-nsvv2    0/1           18s        18s
azure-devops-runners-g82f9   0/1                      4s

kubectl get pods -n azure-devops
NAME                              READY   STATUS    RESTARTS   AGE
azure-devops-runner-nsvv2-8dtb2   2/2     Running   0          27s
```
I don't understand why it creates 2 jobs not only one and why only one pod is spawned. I expect 1 job with 1 pod not 2 jobs with 1 pod.

the job config 
```
  pollingInterval: 20                         # Optional. Default: 30 seconds
  successfulJobsHistoryLimit: 0               # Optional. Default: 100. How many completed jobs should be kept.
  failedJobsHistoryLimit: 0                   # Optional. Default: 100. How many failed jobs should be kept.
  triggers:
  - type: azure-pipelines
    metadata:
      poolID: "39"
      organizationURLFromEnv: "AZP_URL"
      personalAccessTokenFromEnv: "AZP_TOKEN"
      targetPipelinesQueueLength: "1"

```

**Pull request merged**